File: static/programs/aws-s3-state-migration-python/migration.py
from typing import Any

import pulumi

COMPONENT_TYPE = "example:storage:VersionedBucket"
OLD_BUCKET_TYPE = "aws:s3/bucketV2:BucketV2"
BUCKET_TYPE = "aws:s3/bucket:Bucket"
VERSIONING_TYPE = "aws:s3/bucketVersioningV2:BucketVersioningV2"


def state_string(state: dict[str, Any], key: str) -> str:
    value = state.get(key)
    if not isinstance(value, str) or not value:
        raise ValueError(f"State resource has missing or invalid {key}")
    return value


def migrate_bucket_state(
    args: pulumi.StateMigrationArgs,
) -> pulumi.StateMigrationResult | None:
    if not args.old_state:
        return None
    root = args.old_state[0]
    if root.get("type") != COMPONENT_TYPE:
        raise ValueError("Expected a VersionedBucket migration root")

    # Find the bucket and versioning sidecar directly parented to this component.
    index, sidecar_index = -1, -1
    for i, state in enumerate(args.old_state):
        if state.get("parent") != root.get("urn"):
            continue
        if state.get("type") in (OLD_BUCKET_TYPE, BUCKET_TYPE):
            if index != -1:
                raise ValueError("Expected only one bucket")
            index = i
        elif state.get("type") == VERSIONING_TYPE:
            if sidecar_index != -1:
                raise ValueError("Expected only one versioning sidecar")
            sidecar_index = i
    # Leave fresh or already-migrated state unchanged.
    if sidecar_index == -1 and (
        index == -1 or args.old_state[index].get("type") == BUCKET_TYPE
    ):
        return None
    if index == -1 or sidecar_index == -1 or args.old_state[index].get("type") != OLD_BUCKET_TYPE:
        raise ValueError("Expected BucketV2 and its versioning sidecar")
    old_bucket, sidecar = args.old_state[index], args.old_state[sidecar_index]

    # Both state entries must refer to the same physical bucket and provider.
    for key in ("id", "provider"):
        if sidecar.get(key) != state_string(old_bucket, key):
            raise ValueError(f"Bucket and sidecar must have the same {key}")
    # Only migrate the versioning configuration supported by this example.
    outputs = sidecar.get("outputs")
    configuration = outputs.get("versioningConfiguration") if isinstance(outputs, dict) else None
    if (
        not isinstance(configuration, dict)
        or configuration.get("status") != "Enabled"
        or configuration.get("mfaDelete") != "Disabled"
    ):
        raise ValueError("This example requires enabled versioning without MFA delete")
    sidecar_urn = state_string(sidecar, "urn")

    # Copy the bucket state and change its logical type in both the URN and
    # type field, preserving its physical ID and other metadata.
    bucket = dict(old_bucket)
    old_urn = state_string(bucket, "urn")
    old_type = f"${OLD_BUCKET_TYPE}::"
    if old_type not in old_urn:
        raise ValueError("Bucket URN does not match BucketV2")
    new_urn = old_urn.replace(old_type, f"${BUCKET_TYPE}::", 1)
    bucket["urn"], bucket["type"] = new_urn, BUCKET_TYPE

    # Translate both saved inputs and observed outputs to the new schema.
    for field in ("inputs", "outputs"):
        old_properties = bucket.get(field)
        if not isinstance(old_properties, dict):
            raise ValueError(f"Expected bucket {field} object")
        properties = dict(old_properties)
        properties.pop("__pulumi_raw_state_delta", None)
        properties.pop("versionings", None)
        # Populate the new bucket's inline versioning settings from the old
        # versioning sidecar, which managed this configuration before migration.
        properties["versioning"] = {
            "enabled": configuration["status"] == "Enabled",
            "mfaDelete": configuration["mfaDelete"] == "Enabled",
        }

        # BucketV2 uses plural singleton lists; Bucket uses singular objects.
        for plural, singular in {
            "loggings": "logging",
            "replicationConfigurations": "replicationConfiguration",
            "serverSideEncryptionConfigurations": "serverSideEncryptionConfiguration",
            "websites": "website",
        }.items():
            blocks = properties.pop(plural, None)
            if blocks is None:
                continue
            if not isinstance(blocks, list) or len(blocks) > 1:
                raise ValueError(f"Expected at most one {plural} block")
            if blocks:
                properties[singular] = blocks[0]
        bucket[field] = properties

    # Preserve the rest of the subtree and remove the separate versioning entry.
    # Map both old URNs to the new bucket so Pulumi can rewrite references to them.
    new_state = list(args.old_state)
    new_state[index] = bucket
    del new_state[sidecar_index]
    return pulumi.StateMigrationResult(
        new_state=new_state,
        successors={old_urn: new_urn, sidecar_urn: new_urn},
    )

File: static/programs/aws-s3-state-migration-go/migration.go
package main

import (
	"context"
	"fmt"
	"maps"
	"slices"
	"strings"

	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

const (
	componentType  = "example:storage:VersionedBucket"
	oldBucketType  = "aws:s3/bucketV2:BucketV2"
	bucketType     = "aws:s3/bucket:Bucket"
	versioningType = "aws:s3/bucketVersioningV2:BucketVersioningV2"
)

func stateString(state map[string]any, key string) (string, error) {
	value, ok := state[key].(string)
	if !ok || value == "" {
		return "", fmt.Errorf("state resource has missing or invalid %s", key)
	}
	return value, nil
}

func migrateBucketState(_ context.Context, args *pulumi.StateMigrationArgs) (*pulumi.StateMigrationResult, error) {
	if len(args.OldState) == 0 {
		return nil, nil
	}
	root := args.OldState[0]
	if root["type"] != componentType {
		return nil, fmt.Errorf("expected a VersionedBucket migration root")
	}

	// Find the bucket and versioning sidecar directly parented to this component.
	index, sidecarIndex := -1, -1
	for i, state := range args.OldState {
		if state["parent"] != root["urn"] {
			continue
		}
		switch state["type"] {
		case oldBucketType, bucketType:
			if index != -1 {
				return nil, fmt.Errorf("expected only one bucket")
			}
			index = i
		case versioningType:
			if sidecarIndex != -1 {
				return nil, fmt.Errorf("expected only one versioning sidecar")
			}
			sidecarIndex = i
		}
	}
	// Leave fresh or already-migrated state unchanged.
	if sidecarIndex == -1 && (index == -1 || args.OldState[index]["type"] == bucketType) {
		return nil, nil
	}
	if index == -1 || sidecarIndex == -1 || args.OldState[index]["type"] != oldBucketType {
		return nil, fmt.Errorf("expected BucketV2 and its versioning sidecar")
	}
	oldBucket, sidecar := args.OldState[index], args.OldState[sidecarIndex]
	// Both state entries must refer to the same physical bucket and provider.
	for _, key := range []string{"id", "provider"} {
		value, err := stateString(oldBucket, key)
		if err != nil {
			return nil, err
		}
		if sidecar[key] != value {
			return nil, fmt.Errorf("bucket and sidecar must have the same %s", key)
		}
	}
	// Only migrate the versioning configuration supported by this example.
	outputs, _ := sidecar["outputs"].(map[string]any)
	configuration, _ := outputs["versioningConfiguration"].(map[string]any)
	if configuration["status"] != "Enabled" || configuration["mfaDelete"] != "Disabled" {
		return nil, fmt.Errorf("this example requires enabled versioning without MFA delete")
	}
	sidecarURN, err := stateString(sidecar, "urn")
	if err != nil {
		return nil, err
	}

	// Copy the bucket state and change its logical type in both the URN and
	// type field, preserving its physical ID and other metadata.
	bucket := maps.Clone(args.OldState[index])
	oldURN, err := stateString(bucket, "urn")
	if err != nil {
		return nil, err
	}
	oldType := "$" + oldBucketType + "::"
	if !strings.Contains(oldURN, oldType) {
		return nil, fmt.Errorf("bucket URN does not match BucketV2")
	}
	newURN := strings.Replace(oldURN, oldType, "$"+bucketType+"::", 1)
	bucket["urn"], bucket["type"] = newURN, bucketType

	// Translate both saved inputs and observed outputs to the new schema.
	for _, field := range []string{"inputs", "outputs"} {
		oldProperties, ok := bucket[field].(map[string]any)
		if !ok {
			return nil, fmt.Errorf("expected bucket %s object", field)
		}
		properties := maps.Clone(oldProperties)
		delete(properties, "__pulumi_raw_state_delta")
		delete(properties, "versionings")
		// Populate the new bucket's inline versioning settings from the old
		// versioning sidecar, which managed this configuration before migration.
		properties["versioning"] = map[string]any{
			"enabled":   configuration["status"] == "Enabled",
			"mfaDelete": configuration["mfaDelete"] == "Enabled",
		}

		// BucketV2 uses plural singleton lists; Bucket uses singular objects.
		for plural, singular := range map[string]string{
			"loggings":                           "logging",
			"replicationConfigurations":          "replicationConfiguration",
			"serverSideEncryptionConfigurations": "serverSideEncryptionConfiguration",
			"websites":                           "website",
		} {
			value := properties[plural]
			delete(properties, plural)
			if value == nil {
				continue
			}
			blocks, ok := value.([]any)
			if !ok || len(blocks) > 1 {
				return nil, fmt.Errorf("expected at most one %s block", plural)
			}
			if len(blocks) == 1 {
				properties[singular] = blocks[0]
			}
		}
		bucket[field] = properties
	}

	// Preserve the rest of the subtree and remove the separate versioning entry.
	// Map both old URNs to the new bucket so Pulumi can rewrite references to them.
	newState := slices.Clone(args.OldState)
	newState[index] = bucket
	newState = slices.Delete(newState, sidecarIndex, sidecarIndex+1)
	return &pulumi.StateMigrationResult{
		NewState:   newState,
		Successors: map[string]string{oldURN: newURN, sidecarURN: newURN},
	}, nil
}
